ipam/ipv6Ipam.go

```diff
@@ -0,0 +1,281 @@
+	} else if comparison == comparisonError {
+		return nil, errors.New("Error comparing Kubernetes API versions")
+	}
+
+	return client, err
+}
+
+// CompareK8sVer compares two k8s versions.
+// returns -1, 0, 1 if firstVer smaller, equals, bigger than secondVer respectively.
+// returns -2 for error.
+func CompareK8sVer(firstVer *version.Info, secondVer *version.Info) int {
+	v1Minor := re.FindAllString(firstVer.Minor, -1)
+	if len(v1Minor) < 1 {
+		return comparisonError
+	}
+
+	v1, err := semver.NewVersion(firstVer.Major + "." + v1Minor[0])
+	if err != nil {
+		return comparisonError
+	}
+
+	v2Minor := re.FindAllString(secondVer.Minor, -1)
+	if len(v2Minor) < 1 {
+		return comparisonError
+	}
+
+	v2, err := semver.NewVersion(secondVer.Major + "." + v2Minor[0])
+	if err != nil {
+		return comparisonError
+	}
+
+	return v1.Compare(v2)
+}
+
+func (source *ipv6IpamSource) refresh() error {
+	if source == nil {
+		return errors.New("ipv6ipam is nil")
+	}
+
+	if source.subnetRetrieved {
+		return nil
+	}
+
+	if source.kubeClient == nil {
+		kubeClient, err := source.loadKubernetesConfig()
+		if err != nil {
+			return err
+		}
+
+		source.kubeClient = kubeClient
+	}
+
+	kubeNode, err := source.kubeClient.CoreV1().Nodes().Get(context.TODO(), source.nodeHostname, metav1.GetOptions{})
+	if err != nil {
+		log.Printf("[ipam] Failed to retrieve node using hostname: %+v", source.nodeHostname)
+		return err
+	}
+
+	source.kubeNode = kubeNode
+	log.Printf("[ipam] Discovered CIDR's %v.", source.kubeNode.Spec.PodCIDRs)
+
+	// Query the list of Kubernetes Pod IPs
+	interfaceIPs, err := retrieveKubernetesPodIPs(source.kubeNode, source.subnetMaskSizeLimit)
+
+	// Configure the local default address space.
+	local, err := source.sink.newAddressSpace(LocalDefaultAddressSpaceId, LocalScope)
+	if err != nil {
+		return err
+	}
+
+	for _, i := range interfaceIPs.Interfaces {
+		for _, s := range i.IPSubnets {
+			_, subnet, err := net.ParseCIDR(s.Prefix)
+			ifaceName := ""
+			priority := 0
+			ap, err := local.newAddressPool(ifaceName, priority, subnet)
+
+			for _, a := range s.IPAddresses {
+				address := net.ParseIP(a.Address)
+				_, err = ap.newAddressRecord(&address)
+				if err != nil {
+					log.Printf("[ipam] Failed to create address:%v err:%v.", address, err)
+					continue
+				}
+			}
+		}
+	}
+
+	// Set the local address space as active.
+	if err = source.sink.setAddressSpace(local); err != nil {
+		return err
+	}
+
+	source.subnetRetrieved = true
+	log.Printf("[ipam] Address space successfully populated from Kubernetes API Server")
+
+	return err
+}
+
+// retrieves the allocated pod IP's, and populates the NetworkInterfaces struture
+func retrieveKubernetesPodIPs(node *v1.Node, subnetMaskBitSize string) (*NetworkInterfaces, error) {
+	var nodeCidr net.IP
+	var ipnetv6 *net.IPNet
+
+	// get IPv6 subnet allocated to node
+	for _, cidr := range node.Spec.PodCIDRs {
+		nodeCidr, _, _ = net.ParseCIDR(cidr)
+		if nodeCidr.To4() == nil {
+			break
+		}
+	}
+
```

Comment:
check for a case where node doesn't have ipv6 cidr and return err

Reply:
Updated